lib/dry/system/config/component_dirs.rb

```diff
@@ -1,49 +1,181 @@
 require "concurrent/map"
 require "dry/configurable"
+require "dry/system/constants"
 require "dry/system/errors"
 require_relative "component_dir"
 
 module Dry
   module System
     module Config
       class ComponentDirs
-        attr_reader :dirs
+        include Dry::Configurable
 
+        # Settings from ComponentDir can be configured here to apply all added dirs as
+        # defaults
+        @_settings = ComponentDir._settings.dup
```

Comment:
@timriley why is this an ivar?

Reply:
This is a class-level ivar, and it's where dry-configurable stores its setting definitions. By assigning it here, it lets us copy over the settings that are already defined for each individual `ComponentDir` and allow them to be used as defaults.

Another option could've been to create a module that contains the `setting` definitions and mix it into both places, but with the way things are set up, it IMO felt better for `ComponentDir` to "own" those settings.

What do you think?

Reply:
@timriley let's leave it like that but my understanding is that this is a missing feature in dry-configurable, right? If yes, then it's something to improve there so that we could simplify things here 🙂

Reply:
Now that I'm looking at this again a month later(!), I definitely agree that it'd good to have this handled in a nicer way. Ideally the dry-rb gems wouldn't require deep knowledge of the internals of _other_ dry-rb gems to do their job.

The thing is, this _could_ be easily done in a much cleaner way, like I suggested above, where we have a module that we mix into both `Config::ComponentDir` and `Config::ComponentDirs` that defines the settings in each. 

However, if I put the API documentation for the setting in that module, it wouldn't make sense for both including contexts (where one is about configuring individual dirs, and the other about configuring the defaults across _all_ dirs). I think that was pretty much what pushed me in this direction, to have `Config::ComponentDir` "own" the canonical setting definitions, and `Config::ComponentDirs` to copy them over for use when configuring defaults. This also let me define the API documentation in each place, with appropriate wording for each context.

The other thing I had in mind when doing this was what I might need to do if e.g. individual `ComponentDir` instances had a setting that I _didn't_ want configurable globally. In that case, it'd be simple enough to adjust this part here to loop over `ComponentDir._settings` and use an allow/deny list to determine which settings to include/skip when applying here. I'm probably overthinking it, though.

Perhaps one way in of avoiding the nasty class-level ivar assignment here would be to go right ahead do something like I just suggested:

```ruby
ComponentDir._settings.each do |setting|
  _settings << setting.dup
end
```

This is effectively the same as what I'm achieving with `@_settings = ComponentDir._settings.dup` but it does have the benefit of not breaking encapsulation and dealing with the underling ivar directly.

I'm inclined to go ahead and do this. What do you think, @solnic?

Reply:
If we wanted a nicer public API for importing a whole range of settings from another place, I'd advocate for something like `Settings#replace`, which we could use in this context like so:

```ruby
_settings.replace(ComponentDir._settings.dup)
```

`#replace` has precedence across a range of classes within the Ruby stdlib, and since `Settings` is an enumerable collection, having this method here wouldn't be surprising.

Reply:
@timriley thanks for the insights! I love the idea of using `#replace`.

Reply:
I've now updated this to use the following:

```ruby
ComponentDir._settings.each do |setting|
  _settings << setting.dup
end
```

Will file another issue about adding `Settings#replace` to dry-configurable :)